### PR TITLE
Bug fix/make aql fast again 3

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Slightly more efficient handling of SharedAqlItemBlockPtrs

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This change is already covered by existing tests, such as *shell_server_aql*.

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10279/

### CODE REVIEW

File: arangod/Aql/SingleRowFetcher.cpp [hunk 1]
```diff
@@ -209,16 +207,14 @@ template <BlockPassthrough blockPassthrough>
 std::pair<ExecutionState, size_t> SingleRowFetcher<blockPassthrough>::preFetchNumberOfRows(size_t atMost) {
   if (_upstreamState != ExecutionState::DONE && !indexIsValid()) {
     // We have exhausted the current block and need to fetch a new one
-    ExecutionState state;
-    SharedAqlItemBlockPtr newBlock;
-    std::tie(state, newBlock) = fetchBlock(atMost);
+    std::pair<ExecutionState, SharedAqlItemBlockPtr> res = fetchBlock(atMost);
```
Comment (markuspf) [inline arangod/Aql/SingleRowFetcher.cpp:210]:
```suggestion
   auto [state, newBlock] = fetchBlock(atMost);
```

File: arangod/Aql/SingleRowFetcher.cpp [hunk 2]
```diff
@@ -209,16 +207,14 @@ template <BlockPassthrough blockPassthrough>
...
     // we de not need result as local members are modified
-    if (state == ExecutionState::WAITING) {
-      return {state, 0};
+    if (res.first == ExecutionState::WAITING) {
+      return {res.first, 0};
```
Comment (markuspf) [inline arangod/Aql/SingleRowFetcher.cpp:213]:
```suggestion
    if (state == ExecutionState::WAITING) {
      return {state, 0};
```

File: arangod/Aql/SingleRowFetcher.cpp [hunk 3]
```diff
@@ -209,16 +207,14 @@ template <BlockPassthrough blockPassthrough>
...
     }
     // The internal state should be in-line with the returned state.
-    TRI_ASSERT(_upstreamState == state);
-    _currentBlock = std::move(newBlock);
+    TRI_ASSERT(_upstreamState == res.first);
+    _currentBlock = std::move(res.second);
```
Comment (markuspf) [inline arangod/Aql/SingleRowFetcher.cpp:217]:
```suggestion
    TRI_ASSERT(_upstreamState == state);
    _currentBlock = std::move(newBlock);
```

File: arangod/Aql/SingleRowFetcher.cpp (hunk 1 above)
Comment (markuspf) [reply arangod/Aql/SingleRowFetcher.cpp:210]: This is (in my opinion) more readable (not necessarily *here* but further on.

File: arangod/Aql/SingleRowFetcher.cpp [hunk 4]
```diff
@@ -104,14 +104,12 @@ bool SingleRowFetcher<passBlocksThrough>::fetchBlockIfNecessary(size_t atMost) {
     // new one, so we might reuse it immediately!
     _currentBlock = nullptr;
 
-    ExecutionState state;
-    SharedAqlItemBlockPtr newBlock;
-    std::tie(state, newBlock) = fetchBlock(atMost);
-    if (state == ExecutionState::WAITING) {
+    std::pair<ExecutionState, SharedAqlItemBlockPtr> res = fetchBlock(atMost);
+    if (res.first == ExecutionState::WAITING) {
```
Comment (markuspf) [inline arangod/Aql/SingleRowFetcher.cpp:108]:
```suggestion
    auto [state, newBlock] = fetchBlock(atMost);
    if (state == ExecutionState::WAITING) {
```

File: arangod/Aql/SingleRowFetcher.cpp [hunk 5]
```diff
@@ -104,14 +104,12 @@ bool SingleRowFetcher<passBlocksThrough>::fetchBlockIfNecessary(size_t atMost) {
...
       return false;
     }
 
-    _currentBlock = std::move(newBlock);
+    _currentBlock = std::move(res.second);
```
Comment (markuspf) [inline arangod/Aql/SingleRowFetcher.cpp:112]:
```suggestion
    _currentBlock = std::move(newBlock);
```

File: arangod/Aql/ConstFetcher.cpp [hunk 6]
```diff
@@ -148,8 +148,7 @@ auto ConstFetcher::execute(AqlCallStack& stack)
   if (canUseFullBlock(sliceIndexes)) {
     // FastPath
     // No need for slicing
-    SharedAqlItemBlockPtr resultBlock = _blockForPassThrough;
-    _blockForPassThrough.reset(nullptr);
+    SharedAqlItemBlockPtr resultBlock = std::move(_blockForPassThrough);
```
Comment (markuspf) [inline arangod/Aql/ConstFetcher.cpp:151]: could be moved into the return further down as well.

File: arangod/Aql/DistributeClientBlock.cpp [hunk 7]
```diff
@@ -34,7 +34,7 @@ using namespace arangodb::aql;
 DistributeClientBlock::QueueEntry::QueueEntry(SkipResult const& skipped,
                                               SharedAqlItemBlockPtr block,
                                               std::vector<size_t> choosen)
-    : _skip(skipped), _block(block), _choosen(std::move(choosen)) {
+    : _skip(skipped), _block(std::move(block)), _choosen(std::move(choosen)) {
   TRI_ASSERT(_block != nullptr || choosen.empty());
```
Comment (markuspf) [inline arangod/Aql/DistributeClientBlock.cpp:38]: shouldn't this be `_choosen` in the assert? (I know this PR doesn't touch that bit, but anyway)

Comment (markuspf) [reply arangod/Aql/DistributeClientBlock.cpp:38]: it also slightly grates that this should *probably* be called `_chosen`.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10296/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/job/arangodb-matrix-pr-linux/12040/

**Pull request merged**